File: results/statisticalAnalysis.m
%% read log text file
clear all
close all

D = readtable('FullBenchmarkResults.txt','Delimiter',' ');

%% work on positive tests 
close all

DOK=D(D.Var1 > 0,:);
testNumber = size(D,1);
sucessNumber = size(DOK,1);
fprintf('Probability of Sucess : %d (%d/%d) \n',sucessNumber/testNumber,sucessNumber,testNumber);

figure()
edges = [min(DOK.Var3)-1:1:max(DOK.Var3)+1]; 
H = histogram(DOK.Var3,edges);
title('Sucessfull tests : Histogram of the Travelled Distance')
fprintf('Secondary Indicators Format : "Name : Mean [First Decile, Ninth Decile]"\n')
fprintf('Travelled Distance (m) : %d [%d, %d] \n',quantile(DOK.Var3,0.5),quantile(DOK.Var3,0.1),quantile(DOK.Var3,0.9));


figure()
edges = [min(DOK.Var4)-1:0.5:max(DOK.Var4)+1]; 
H = histogram(DOK.Var4,edges);
title('Sucessfull tests : Histogram of the Completion Time')
fprintf('Time to complete the tet (s) : %d [%d, %d] \n',quantile(DOK.Var4,0.5),quantile(DOK.Var4,0.1),quantile(DOK.Var4,0.9));

figure()
edges = [min(DOK.Var5)-1:0.5:max(DOK.Var5)+1]; 
H = histogram(DOK.Var5,edges);
title('Sucessfull tests : Histogram of the Consumed Energy')
fprintf('Consumed Energy (Wh) : %d [%d, %d] \n',quantile(DOK.Var5,0.5),quantile(DOK.Var5,0.1),quantile(DOK.Var5,0.9));

figure()
averageSpeed = DOK.Var3./DOK.Var4;
edges = [min(averageSpeed)-0.1:0.1:max(averageSpeed)+.1]; 
H = histogram(averageSpeed,edges);
title('Sucessfull tests : Histogram of the flying Speed')
fprintf('Average Speed (m/s) : %d [%d, %d] \n',quantile(averageSpeed,0.5),quantile(averageSpeed,0.1),quantile(averageSpeed,0.9));


%% work on non finished tests
DFAILED = D(D.Var1 < 1,:);

%Print number of tests which are not finished after 60s, some of them are
%bugged, I want to erase those. 
failNumber = size(DFAILED,1);


figure()
edges = [min(DFAILED.Var4)-1:1:max(DFAILED.Var4)+1]; 
H = histogram(DFAILED.Var4,edges);
title('Failed test : Histogram of the Flight Length Before Collision')
fprintf('Failed tests : Flight Length Before Collision (s) : %d [%d, %d] \n',quantile(DFAILED.Var4,0.5),quantile(DFAILED.Var4,0.1),quantile(DFAILED.Var4,0.9));

figure()
edges = [min(DFAILED.Var6)-1:1:max(DFAILED.Var6)+1]; 
H = histogram(DFAILED.Var6,edges);
title('Failed test : Histogram of the Linear Distance Before Collision')
fprintf('Failed tests : Linear Distance Before Collision (m) : %d [%d, %d] \n',quantile(DFAILED.Var6,0.5),quantile(DFAILED.Var6,0.1),quantile(DFAILED.Var6,0.9));

%% save the table or get info on the table
% FileName=[datestr(now, 'yyyy-mm-dd_HH:MM:SS'),'_benchmarkAnalysis.dat'];
% save(FileName,'D') ;
 

File: results/statisticalAnalysis.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Very light (numpy not needed) script that compute the benchmark statistical indicators."""

import sys
import math

print("")
print("STATISTICAL ANALYSIS:")
print("")

with open(sys.argv[1]) as textFile:
    data_ligne = [line.split() for line in textFile] #initial storing of all results in data_ligne

    # create two array data_success and data_failed that store the results as floats,
    # tranpose the initial data and store only the results when the test pass
    # compute the sucess rate simultaneously
    # and compute the average flying speed
    data_success = [[] for _ in range(len(data_ligne[0])+1)]
    data_failed = []
    sucessNumber = 0;
    for i in range(len(data_ligne)):
        sucessNumber += float(data_ligne[i][0])
        if float(data_ligne[i][0]) == 1:
            for j in range(len(data_ligne[i])):
                data_success[j].append(float(data_ligne[i][j]))
            data_success[len(data_ligne[i])].append(data_success[2][len(data_success[2])-1]/data_success[3][len(data_success[3])-1])
        if float(data_ligne[i][0]) == 0:
                data_failed.append(float(data_ligne[i][len(data_ligne[i])-1]))
    
    if sucessNumber == 0:
        print("ONLY UNSUCESSFULL FLIGHTS OCCURED.")
        print("The secondary indicators canno't be computed in this case.")
        sys.exit("")

    # compute and store the mean of all indicators
    means = [0.0] * 8
    means[0] = sucessNumber/float(len(data_ligne))
    for i in range(1, 7):
        means[i] = sum(data_success[i])/float(len(data_success[i]))
    means[7] = sum(data_failed)/float(len(data_failed))
        
    # compute the 99% confidance interval on the collision probability (see chernoff bound) 
    estError = math.sqrt(math.log1p(2/0.01)/(2*len(data_ligne)))
    lowsucessbound = max(0, means[0] - estError)
    highsucessbound = min(1, means[0] + estError)

    # Sort the data and compute the decile Ids for both sucess and failed data vectors
    for indic in data_success:
        indic.sort()
    idFirstDecile = int(len(data_success[0]) / 10 )
    idLastDecile = int(len(data_success[0]) * 9 / 10)

    data_failed.sort()
    idFailedFirstDecile = int(len(data_failed) / 10 )
    idFailedLastDecile = int(len(data_failed) * 9 / 10)


    print("STATISTICAL SUCESS RATE: %.2f" % means[0])
    print("Over the %i tests you performed, it means the probablity of sucess is in [%.2f, %.2f] with a 99%% condidence"  %(len(data_ligne), lowsucessbound, highsucessbound))
    print("Secondary Indicators Format : 'Name : Mean [First Decile, Ninth Decile]'")
    print("Travelled Distance (m) : %.2f [%.2f, %.2f]" %(means[2], data_success[2][idFirstDecile], data_success[2][idLastDecile] ))
    print("Time to complete the Test (s) : %.2f [%.2f, %.2f]" %(means[3], data_success[3][idFirstDecile], data_success[3][idLastDecile]))
    print("Consumed Energy (Wh) : %.2f [%.2f, %.2f]" %(means[4], data_success[4][idFirstDecile], data_success[4][idLastDecile]))
    print("Average Speed (m/s) : %.2f [%.2f, %.2f]" %(means[len(data_ligne[0])], data_success[len(data_ligne[0])][idFirstDecile], data_success[len(data_ligne[0])][idLastDecile]))
    print("Average Linear Distance before collision over failed Tests (m) : %.2f [%.2f, %.2f]" %(means[7], data_failed[idFailedFirstDecile], data_failed[idFailedLastDecile]))
